### PR TITLE
Started adding ESP32-H2 support

### PR DESCRIPTION
**Pull request by playfulFence**
## clocks
   - Implemented all functionality required to get `low-level-clocks` working
      -  added `esp-hal-common/src/clock/clock_ll/esp32h2`, modified `clock/mod.rs`
## esp32h2-hal
   - created `esp32h2-hal` package
      - updated linker scripts
      - :exclamation: need to update `esp-backtrace` package link in `Cargo.toml` after it's merged 
   - :exclamation: README.md needs to be updated

### CODE REVIEW

**Review by jessebraham — CHANGES_REQUESTED** [review]
Could you please add `esp32h2-hal` to the CI workflow and update `Cargo.toml`?

File: esp32h2-hal/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+name              = "embassy_hello_world"
+required-features = ["embassy"]
+
+[[example]]
+name              = "embassy_wait"
+required-features = ["embassy", "async"]
+
+[[example]]
+name              = "embassy_spi"
+required-features = ["embassy", "async"]
```
[the remark this answers is not in the document]
Comment (playfulFence) [reply esp32h2-hal/Cargo.toml]: > These will need to be either commented out or removed until the examples are added.

`Cargo.toml` has been updated (also the `esp-backtrace` has been replaced with the official one with merged updates)

[the remark this answers is not in the document]
Comment (playfulFence) [reply esp32h2-hal/Cargo.toml]: `esp32h2-hal` and all `esp32h2`-related (possible at the moment) things were also added to CI workflow

File: esp-hal-common/src/clock/mod.rs [hunk 2]
```diff
@@ -429,6 +434,58 @@ impl<'d> ClockControl<'d> {
...
+    pub fn boot_defaults(
+        clock_control: impl Peripheral<P = SystemClockControl> + 'd,
+    ) -> ClockControl<'d> {
+        ClockControl {
+            _private: clock_control.into_ref(),
+            desired_rates: RawClocks {
+                cpu_clock: HertzU32::MHz(96),  //
+                apb_clock: HertzU32::MHz(32),  //
+                xtal_clock: HertzU32::MHz(32), //
+                i2c_clock: HertzU32::MHz(32),  //
```
Comment (jessebraham) [inline esp-hal-common/src/clock/mod.rs:450]: Sorry to nitpick, but let's get rid of these trailing comments please :)

File: esp-hal-common/src/clock/mod.rs [hunk 3]
```diff
@@ -429,6 +434,58 @@ impl<'d> ClockControl<'d> {
...
+            clocks_ll::esp32h2_rtc_apb_freq_update(apb_freq);
+        }
+
+        ClockControl {
+            _private: clock_control.into_ref(),
+            desired_rates: RawClocks {
+                cpu_clock: cpu_clock_speed.frequency(),
+                apb_clock: apb_freq.frequency(),
+                xtal_clock: xtal_freq.frequency(),
+                i2c_clock: HertzU32::MHz(96),
```
Comment (jessebraham) [inline esp-hal-common/src/clock/mod.rs:483]: This doesn't match the value defined above for `i2c_clock`, is this correct?

File: esp-hal-common/src/rtc_cntl/mod.rs [hunk 4]
```diff
@@ -754,20 +759,21 @@ pub fn get_reset_reason(cpu: Cpu) -> Option<SocResetReason> {
...
     // FIXME: check s_light_sleep_wakeup
-    // https://github.com/espressif/esp-idf/blob/afbdb0f3ef195ab51690a64e22bfb8a5cd487914/components/esp_hw_support/sleep_modes.c#L1394
-    if get_reset_reason(Cpu::ProCpu).unwrap() != SocResetReason::CoreDeepSleep {
-        return SleepSource::Undefined;
-    }
+    // https://github.com/espressif/esp-idf/blob/afbdb0f/components/esp_hw_support/sleep_modes.c#L1394
+    // FIXME: un-comment block when H2's `SocResetReason` enum is populated
+    // if get_reset_reason(Cpu::ProCpu).unwrap() != SocResetReason::CoreDeepSleep {
+    //     return SleepSource::Undefined;
+    // }
```
Comment (jessebraham) [inline esp-hal-common/src/rtc_cntl/mod.rs:766]: If possible lets populate this enum please; if it's a hassle we can wait, but I'd rather just get it done.

File: esp-hal-common/src/soc/esp32h2/mod.rs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+pub mod efuse;
+pub mod gpio;
+pub mod peripherals;
+pub mod radio_clocks;
+
+pub(crate) mod registers {
+    pub const INTERRUPT_MAP_BASE: u32 = 0x60010000; // FIXME
```
Comment (jessebraham) [inline esp-hal-common/src/soc/esp32h2/mod.rs:7]: Let's fix this please!

File: esp-hal-common/src/rtc_cntl/mod.rs [hunk 6]
```diff
@@ -606,21 +607,25 @@ impl Rwdt {
...
         let rtc_cntl = unsafe { &*RTC_CNTL::PTR };
-        #[cfg(esp32c6)]
+        #[cfg(any(esp32c6, esp32h2))]
         let rtc_cntl = unsafe { &*LP_WDT::PTR };
+
         let wkey = if enable { 0u32 } else { 0x50D8_3AA1 };
 
-        rtc_cntl.wdtwprotect.write(|w| unsafe { w.bits(wkey) });
+        // FIXME: register missing for H2
+        // rtc_cntl.wdtwprotect.write(|w| unsafe { w.bits(wkey) });
```
[the remark this answers is not in the document]
Comment (jessebraham) [reply esp-hal-common/src/rtc_cntl/mod.rs:618]: I will fix the PAC in a bit here and we can un-comment this line

File: esp-hal-common/src/soc/esp32h2/mod.rs (hunk 5 above)
Comment (playfulFence) [reply esp-hal-common/src/soc/esp32h2/mod.rs]: Comment deleted, value is already correct

File: esp-hal-common/Cargo.toml [hunk 7]
```diff
@@ -54,6 +54,7 @@ esp32   = { version = "0.23.0", features = ["critical-section"], optional = true
 esp32c2 = { version = "0.10.0", features = ["critical-section"], optional = true }
 esp32c3 = { version = "0.13.0", features = ["critical-section"], optional = true }
 esp32c6 = { version = "0.3.0",  features = ["critical-section"], optional = true }
+esp32h2 = { git = "https://github.com/esp-rs/esp-pacs", rev = "3db0042", package = "esp32h2", features = ["critical-section"], optional = true }
```
Comment (jessebraham) [inline esp-hal-common/Cargo.toml:57]:
```suggestion
esp32h2 = { git = "https://github.com/esp-rs/esp-pacs", rev = "0ee1cea", package = "esp32h2", features = ["critical-section"], optional = true }
```

This new commit revision should fix the `rtc_cntl.wdtwprotect` register you had to comment out.

File: esp-hal-common/src/rtc_cntl/mod.rs (hunk 4 above)
Comment (playfulFence) [reply esp-hal-common/src/rtc_cntl/mod.rs]: Enum was populated and function is uncommented

File: esp-hal-common/src/clock/clocks_ll/esp32h2.rs [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+const REGI2C_RTC_SLAVE_ID_S: u8 = 0;
+const REGI2C_RTC_ADDR_V: u8 = 0xFF;
+const REGI2C_RTC_ADDR_S: u8 = 8;
+const REGI2C_RTC_WR_CNTL_V: u8 = 0x1;
+const REGI2C_RTC_WR_CNTL_S: u8 = 24;
+const REGI2C_RTC_DATA_V: u8 = 0xFF;
+const REGI2C_RTC_DATA_S: u8 = 16;
+
+const I2C_MST_I2C0_CTRL_REG: u32 = DR_REG_I2C_ANA_MST_BASE + 0x0;
+const I2C_ANA_MST_I2C0_BUSY: u32 = 1 << 25;
```
Comment (jessebraham) [inline esp-hal-common/src/clock/clocks_ll/esp32h2.rs:69]: This is not being referenced by anything and can be removed.

**Review by jessebraham — CHANGES_REQUESTED** [review]

File: esp-hal-common/src/rtc_cntl/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (playfulFence) [reply esp-hal-common/src/rtc_cntl/mod.rs]: > I will fix the PAC in a bit here and we can un-comment this line

Done

**Pull request merged**